src/core/inbox.rs

```diff
@@ -1,53 +1,124 @@
-        .dereference(data, local_instance, request_counter)
-        .await?;
-    verify_signature(&request, actor.public_key())?;
+#[cfg(feature = "axum")]
+mod axum_imp {
+    use crate::{
+        core::{object_id::ObjectId, signatures::axum::verify_signature},
+        data::Data,
+        traits::{ActivityHandler, Actor, ApubObject},
+        utils::{verify_domains_match, verify_url_valid},
+        Error,
+        LocalInstance,
+    };
+    use axum::{http::Request, Extension, Json};
+    use serde::de::DeserializeOwned;
+    use tracing::debug;
 
-    debug!("Verifying activity {}", activity.id().to_string());
-    activity.verify(data, request_counter).await?;
+    /// Receive an activity and perform some basic checks, including HTTP signature verification.
+    pub async fn receive_activity<Activity, ActorT, Datatype>(
+        request: Request<Activity>,
+        Json(activity): Json<Activity>,
+        Extension(local_instance): Extension<&LocalInstance>,
+        Extension(data): Extension<&Data<Datatype>>,
+    ) -> Result<(), <Activity as ActivityHandler>::Error>
+    where
+        Activity: ActivityHandler<DataType = Datatype> + DeserializeOwned + Send + 'static,
+        ActorT: ApubObject<DataType = Datatype> + Actor + Send + 'static,
+        for<'de2> <ActorT as ApubObject>::ApubType: serde::Deserialize<'de2>,
+        <Activity as ActivityHandler>::Error: From<anyhow::Error>
+            + From<Error>
+            + From<<ActorT as ApubObject>::Error>
+            + From<serde_json::Error>,
+        // FIXME
+        // + From<http_signature_normalization_actix::digest::middleware::VerifyError>,
+        <ActorT as ApubObject>::Error: From<Error> + From<anyhow::Error>,
+    {
+        // ensure that payload hash was checked against digest header by middleware
+        /// FIXME
+        // DigestVerified::from_request(&request, &mut Payload::None).await?;
+        verify_domains_match(activity.id(), activity.actor())?;
+        verify_url_valid(activity.id(), &local_instance.settings).await?;
+        if local_instance.is_local_url(activity.id()) {
+            return Err(
+                Error::UrlVerificationError("Activity was sent from local instance").into(),
+            );
+        }
 
-    debug!("Receiving activity {}", activity.id().to_string());
-    activity.receive(data, request_counter).await?;
-    Ok(HttpResponse::Ok().finish())
+        let request_counter = &mut 0;
+        let actor = ObjectId::<ActorT>::new(activity.actor().clone())
+            .dereference(data, local_instance, request_counter)
+            .await?;
+        verify_signature(&request, actor.public_key())?;
+
+        debug!("Verifying activity {}", activity.id().to_string());
+        activity.verify(data, request_counter).await?;
+
+        debug!("Receiving activity {}", activity.id().to_string());
+        activity.receive(data, request_counter).await?;
+        Ok(())
+    }
```

Comment:
Would be nice if you could extract common functionality into a helper function, so that its not duplicated in two places.